Add support for confounding variables

KochTobi opened the pull request:
Addresses #642

KochTobi commented:
As confounding variables are valid for all samples in an experiment (including samples where a value might not yet been known), I decided to go for including empty variables in the samples.

KochTobi commented:
For this to work, [sql/complete-schema.sql](https://github.com/qbicsoftware/data-manager-app/blob/5b0befc6f99b394c0dd848de4afe01382d684e80/sql/complete-schema.sql) has to be executed on the database

sven1103 commented:
@KochTobi there are some conflicts with the target branch `development`

sven1103 commented:
@KochTobi can you elaborate more on the changes in the database? I also encourage to look into database versioning solutions like Flyway. I have some concerns with our current database schema development strategy: hoping for the best and no way back.

sven1103 commented on project-management/src/main/java/life/qbic/projectmanagement/application/sample/SampleMetadata.java:
```diff
@@ -49,12 +54,25 @@ public static SampleMetadata createUpdate(SampleId sampleId,
         experimentId);
   }
 
+  public static SampleMetadata fromNewToExisting(SampleId id, SampleMetadata sampleMetadata) {
```
I am unsure about the naming here: the method just creates a new sample metadata record with the given sample ID. 
I suggest a naming that describes what it does, instead of assuming context. 

`addSampleID(SampleId id, SampleMetadata metadata)`

sven1103 commented on project-management/src/main/java/life/qbic/projectmanagement/application/sample/SampleRegistrationServiceV2.java:
```diff
@@ -42,32 +51,46 @@ public class SampleRegistrationServiceV2 {
       throws RegistrationException {
     var result = batchRegistrationService.registerBatch(batchLabel, batchIsPilot, projectId,
-        experimentId);
+        ExperimentId.parse(experiment.id()));
```
I think we overdid it here big time with the Identifier... Imo there is no need to parse, because it does not have any semantics or schema. It is just a string value generated from the UUID generator. You either find an object with the ID in the database, or you dont. 
Happy to see some simplification.

KochTobi replied:
I agree!

sven1103 commented on project-management/src/main/java/life/qbic/projectmanagement/application/sample/SampleRegistrationServiceV2.java:
```diff
@@ -42,32 +51,46 @@ public class SampleRegistrationServiceV2 {
 
   @PreAuthorize("hasPermission(#projectId, 'life.qbic.projectmanagement.domain.model.project.Project', 'WRITE')")
   @Async
+  @Transactional(propagation = Propagation.REQUIRES_NEW)
```
What does this annotation do? Why is it needed?

KochTobi replied:
The transactional annotation is needed to indicate a transaction taking place. It starts a transaction in the database and commits changes when exiting the method. The propagation property tells spring that a new transaction is to be created for this operation. The default behaviour is that an existing active transaction is used and if there is none that one is created. I do not think we need to explicity create a new transaction here so I removed the propagation type.

KochTobi replied:
There are different scenarios:

1. no `@Transactional` annotation
2. a `@Transactional` annotation without `Propagation.REQUIRES_NEW`
3. a `@Transactional` annotation the default propagation

in 1. there will be an exception if the method is called from outside a transactional context. In our case no transactional context exists so the exception will be thrown. Without any transaction, no rollback is possible

in 2. at the start of the method a new transaction will be created, suspending an existing transaction if any already exists in this scope. The changes are made within the new transaction which is committed when the method exits (or rolled back in case of exceptions). After the newly created transaction finishes (commit or rollback) the previously suspended transaction continues. 

in 3. when called from a transactional scope (already existing transaction) the operations within the method are appended/added to the existing transaction. If the method throws an exception the complete transaction (including the changes made before this method was called) are rolled back. If no transaction existed when the method is called, a new transaction is created.

The difference between 2. and 3. is that in 2. a transaction is suspended. Rollbacks do not affect a scope outside of the method in 2. Suspending transactions bears a risk of deadlocking the system. If the amount of maximal transactions is limited and the creation of the transaction fails, there might be a scenario where the original transaction is kept in a suspended state further limiting the amount of available transactions. 
We also want to rollback an existing transaction in case of exceptions so only rolling back changes made in this method is not what we want I think.
Thus the propagation type of `REQUIRES_NEW` was wrong here.

sven1103 commented on project-management/src/main/java/life/qbic/projectmanagement/application/sample/SampleRegistrationServiceV2.java:
```diff
@@ -42,32 +51,46 @@ public class SampleRegistrationServiceV2 {
+      var registeredSamples = registerSamples(sampleMetadata, batchId, projectId);
+      for (Entry<Sample, SampleMetadata> registeredSample : registeredSamples.entrySet()) {
+        Map<VariableReference, String> levels = new HashMap<>();
+        registeredSample.getValue().confoundingVariables().forEach((k, v) -> levels.put(k.id(), v));
```
suggestion: you can use `Collectors.toMap()` here, makes it more concise.

KochTobi replied:
To use `Collectors#toMap`, I need to create a stream first. So it does not make it more readable I feel. However I have no strong opinion on the matter. So if you think the lower example is better I will change it.
```
confoundingVariables().foreach((k, v) -> levels.put(k.id(), v));
```
```
registeredSample.getValue().confoundingVariables()
            .entrySet()
            .stream()
            .collect(Collectors.toMap(
                entry -> entry.getKey().id(),
                Entry::getValue));
```

sven1103 replied:
Yes, just nit, but to be fair it is:

```java
Map<VariableReference, String> levels = new HashMap<>();
registeredSample.getValue().confoundingVariables().forEach((k, v) -> levels.put(k.id(), v));
```

vs 

```java
var levels = registeredSample.getValue().confoundingVariables()
                                .entrySet().stream()
                                .collect(Collectors.toMap(entry -> entry.getKey().id(), Entry::getValue));

The last one is side-effect free, vs the first is depending on an external map initialization :P

sven1103 commented on project-management/src/main/java/life/qbic/projectmanagement/application/sample/SampleRegistrationServiceV2.java:
```diff
@@ -101,14 +126,25 @@ public CompletableFuture<Void> updateSamples(
+    for (Entry<Sample, SampleMetadata> updatedSample : updatedSamples.entrySet()) {
+      Map<VariableReference, String> levels = new HashMap<>();
+
+      updatedSample.getValue().confoundingVariables().forEach((k, v) -> levels.put(k.id(), v));
```
Again ``Collectors.toMap()``

KochTobi replied:
see https://github.com/qbicsoftware/data-manager-app/pull/877#discussion_r1930180547

sven1103 reviewed (changes requested):
Some minor change requests, but I have some concerns with how we deal with database changes.

KochTobi commented:
> @KochTobi there are some conflicts with the target branch `development`

Yes I resolved it.

KochTobi commented:
> @KochTobi can you elaborate more on the changes in the database? I also encourage to look into database versioning solutions like Flyway. I have some concerns with our current database schema development strategy: hoping for the best and no way back.

Hi Sven. I simply merged all the sql files to set up the database into one and added the tables for confounding variables. In this case the database schema is simply extended by the tables of confounding variables (samples, experiment and so forth are not touched).
I will happily look into Flyway or other database migration solutions. In the case of this PR I only recorded the current database scheme we have (plus the tables I added).

Even thou the topic of database migration and future proofing is of great importance, I would like to separate the investigation on database migration and future proofing from this PR.

sven1103 reviewed (approved):
Great work @KochTobi, i am looking forward to the users feedback!

KochTobi commented:
> > @KochTobi can you elaborate more on the changes in the database? I also encourage to look into database versioning solutions like Flyway. I have some concerns with our current database schema development strategy: hoping for the best and no way back.
> 
> Hi Sven. I simply merged all the sql files to set up the database into one and added the tables for confounding variables. In this case the database schema is simply extended by the tables of confounding variables (samples, experiment and so forth are not touched). I will happily look into Flyway or other database migration solutions. In the case of this PR I only recorded the current database scheme we have (plus the tables I added).
> 
> Even thou the topic of database migration and future proofing is of great importance, I would like to separate the investigation on database migration and future proofing from this PR.

https://github.com/qbicsoftware/data-manager-app/issues/1000 <- to track the need
